ethers-middleware/src/proxy_wallet/ds_proxy/mod.rs

```diff
@@ -0,0 +1,89 @@
+use super::{ProxyWallet, ProxyWalletError};
+use ethers_contract::BaseContract;
+use ethers_core::{
+    abi::{parse_abi, Tokenize},
+    types::*,
+};
+
+const DS_PROXY_EXECUTE_SIGNATURE: &str =
+    "function execute(address target, bytes data) public payable returns (bytes memory response)";
+
+#[derive(Debug)]
+pub struct DsProxy {
+    address: Address,
+}
+
+impl DsProxy {
+    /// Create a new instance of DsProxy by providing the address of the DsProxy contract that has
+    /// already been deployed to the Ethereum network.
+    pub fn new(address: Address) -> Self {
+        Self { address }
+    }
+}
+
+impl ProxyWallet for DsProxy {
+    fn get_proxy_tx(&self, tx: TransactionRequest) -> Result<TransactionRequest, ProxyWalletError> {
+        // clone the tx into a new proxy tx.
+        let mut proxy_tx = tx.clone();
+
+        // the target address cannot be None.
+        let mut target = Address::default();
+        if let Some(NameOrAddress::Address(addr)) = tx.to {
+            target = addr;
+        } else {
+            return Err(ProxyWalletError::Dummy);
+        }
+
+        // fetch the data field.
+        let data = tx.data.unwrap_or(vec![].into());
+
+        // encode data as the ABI encoded data for DSProxy's execute method.
+        let ds_proxy_base: BaseContract = parse_abi(&[DS_PROXY_EXECUTE_SIGNATURE])?.into();
+        let encoded_data = ds_proxy_base.encode("execute", (target, data))?;
+
+        // update appropriate fields of the proxy tx.
+        proxy_tx.data = Some(encoded_data);
+        proxy_tx.to = Some(NameOrAddress::Address(self.address));
+
+        Ok(proxy_tx)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use hex::FromHex;
+
+    #[test]
+    fn test_get_proxy_tx() {
+        let ds_proxy_addr = Address::random();
+        let ds_proxy = DsProxy::new(ds_proxy_addr);
+
+        let tx: TransactionRequest = serde_json::from_str(
```

Comment:
Where is this data from? A [transaction request](https://github.com/gakonst/ethers-rs/blob/master/ethers-core/src/types/transaction.rs#L22) does not have an`transaction_index` or a `chain_id`

A better way to structure this test could be:

1. create a `BaseContract` with a `set(uint256 x)` function
2. create the calldata for the call by encoding the function data for a sample call
3. call `get_proxy_tx` with the calldata from 2 and a dummy address to get the DsProxy's calldata
4. decode the data from 3 and assert that it matches the calldata from 2 and the dummy address